### PR TITLE
Suggest obvious fix when action id includes type

### LINKED ISSUE #166: Suggested replacement for unrecognized action entity misses obvious fix
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/cedar-policy/cedar/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/cedar-policy/cedar/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Bug Category

Schemas and Validation

### Describe the bug

The `actions` section of the schema lists entity IDs, whereas the `entityTypes` section lists entity types. If a user mistakenly includes the `Action` entity type (and any namespace) as part of their action ID, the error messages are confusing. 

### Expected behavior

Ideally, the error message can identify the user's mistake more clearly, and do better at suggesting a fix. The suggestion would need to account for this particular mistake of including part of the type/namespace in the action ID.

### Reproduction steps

In [tinytodo's `tinytodo.cedarschema.json`](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), modify the `actions` element `"createList"` to instead be `"Action::CreateList"`. When validating `polices.cedar` we get the error message
```
Validation error on policy policy0: Unrecognized action id Action::"CreateList", did you mean Action::"GetList"?
```
This is confusing! With this change, we are defining an action entity `Action::"Action::CreateList"`. This is not what the user intended, because they failed to appreciate the distinction between entity IDs and entity types.

### Code Snippet

_No response_

### Log output

_No response_

### Additional configuration

_No response_

### Operating  System

_No response_

### Additional information and screenshots

_No response_

### PR DESCRIPTION
**Pull request by adpaco-aws**
## Description of changes

This PR extends the validation of action IDs so that we detect cases in which the user may have mistakenly included the entity type (i.e., `"Action::"`) in the action ID. In that case, we don't suggest another action ID like we used to, but instead draw attention to the fact that the action ID was defined with the entity type in its name. This avoids misleading the user when the fix is much simpler.

Resolves #166 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.

### CODE REVIEW

File: cedar-policy-validator/src/rbac.rs [hunk 1]
```diff
@@ -79,7 +78,6 @@ impl Validator {
         let known_action_ids = self
             .schema
             .known_action_ids()
-            .map(ToString::to_string)
             .collect::<Vec<_>>();
```
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/rbac.rs:81]: Might be able to get rid of this `collect` by using `&Schema` as the parameter type for `unrecognized_action_id_help`

File: cedar-policy-validator/src/diagnostics/validation_errors.rs [hunk 2]
```diff
@@ -72,20 +73,55 @@ pub struct UnrecognizedActionId {
...
+/// Help for resolving an unrecognized action id error
+#[derive(Debug, Clone, Error, Hash, Eq, PartialEq)]
+pub enum UnrecognizedActionIdHelp {
+    /// Draw attention to action id including action type (e.g., `Action::"Action::view"`)
+    AvoidActionTypeInActionId(String),
+    /// Suggest an alternative action
+    SuggestAlternative(String),
+}
+
+impl std::fmt::Display for UnrecognizedActionIdHelp {
```
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/diagnostics/validation_errors.rs:100]: Do we need this Display impl? Alternatively, so we just use it in the `help` function?

File: cedar-policy-validator/src/rbac.rs [hunk 3]
```diff
@@ -727,6 +725,38 @@ mod test {
         assert_eq!(notes.len(), 1, "{:?}", notes);
     }
 
+    #[test]
+    fn validate_action_id_with_action_type() {
```
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/rbac.rs:729]: Probably worth adding a test case with a non-empty namespace

**Review by john-h-kastner-aws — APPROVED** [review]
Looks pretty good.

**Review by john-h-kastner-aws — COMMENTED** [review]
IMO worth adding an line in the changelog. We weren't consistent with mentioning error message changes like this leading up to the 4.0 release since we had made a ton of error message tweaks, but I like including these small improvements.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs [hunk 4]
```diff
@@ -72,20 +73,55 @@ pub struct UnrecognizedActionId {
     pub actual_action_id: String,
     /// An action id from the schema that the user might reasonably have
     /// intended to write.
-    pub suggested_action_id: Option<String>,
+    pub unrecognized_action_id_help: Option<UnrecognizedActionIdHelp>,
 }
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/diagnostics/validation_errors.rs:77]: Probably worth updating the doc comment.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy-validator/src/diagnostics/validation_errors.rs:100]: Alternatively, our pattern elsewhere is to not manually impl `Display` for types that `derive(Error)` like this one -- instead, you can use `#[error("blah blah blah")]` on each enum variant to put the `Display` impl inline in the definition.  One example of this is the declaration of `LubHelp` in this same file.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs [hunk 5]
```diff
@@ -72,20 +73,55 @@ pub struct UnrecognizedActionId {
...
+            Self::SuggestAlternative(s) => write!(f, "the action id {s} may be an alternative"),
+        }
+    }
+}
+
+/// Determine the help to offer in the presence of an unrecognized action id error.
+pub fn unrecognized_action_id_help(euid: &EntityUID, euids: &[&EntityUID]) -> Option<UnrecognizedActionIdHelp> {
+    // Check if the user has included the type (i.e., `Action::`) in the action id
+    let eid_with_type = format!("Action::{}", <Eid as AsRef<str>>::as_ref(euid.eid()));
+    let maybe_id_with_type = euids.iter().find(|euid| <Eid as AsRef<str>>::as_ref(euid.eid()).to_string().contains(&eid_with_type));
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/diagnostics/validation_errors.rs:113]: nit: do we need the `.to_string()`? does `&str` not support `.contains()`?

**Review by cdisselkoen — APPROVED** [review]
Looks good overall.

File: cedar-policy-validator/src/rbac.rs (hunk 1 above)
Comment (adpaco-aws) [reply cedar-policy-validator/src/rbac.rs]: Ok, we're now passing `&Schema`

File: cedar-policy-validator/src/rbac.rs (hunk 3 above)
Comment (adpaco-aws) [reply cedar-policy-validator/src/rbac.rs:727]: Added one.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs (hunk 5 above)
Comment (adpaco-aws) [reply cedar-policy-validator/src/diagnostics/validation_errors.rs]: You're right, but for some reason I thought it wasn't. Removed it.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs (hunk 4 above)
Comment (adpaco-aws) [reply cedar-policy-validator/src/diagnostics/validation_errors.rs]: Updated and renamed this field to `hint` as we do in other types.

File: cedar-policy-validator/src/diagnostics/validation_errors.rs [hunk 6]
```diff
@@ -72,20 +73,55 @@ pub struct UnrecognizedActionId {
...
+            Self::AvoidActionTypeInActionId(s) => write!(f, "the action id {s} is similar but includes the type"),
+            Self::SuggestAlternative(s) => write!(f, "the action id {s} may be an alternative"),
+        }
+    }
+}
+
+/// Determine the help to offer in the presence of an unrecognized action id error.
+pub fn unrecognized_action_id_help(euid: &EntityUID, euids: &[&EntityUID]) -> Option<UnrecognizedActionIdHelp> {
+    // Check if the user has included the type (i.e., `Action::`) in the action id
+    let eid_with_type = format!("Action::{}", <Eid as AsRef<str>>::as_ref(euid.eid()));
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply cedar-policy-validator/src/diagnostics/validation_errors.rs]: Yes, there's no reason not to, but I'll admit I don't know how one goes about creating such action ID. Regardless, we're now checking if the entity IDs contain either `Action::<ID>` or `Action::\"<ID>\"`.

File: cedar-policy/CHANGELOG.md [hunk 7]
```diff
@@ -23,6 +23,11 @@ Cedar Language Version: TBD
   Cedar request. To use this API you must enable the `entity-manifest` feature flag.
 - Added public APIs to get language and SDK version numbers (#1219).
 
+### Changed
+
+- The validator provides a more specific hint when an action ID cannot be found
+  and the same action ID with `Action::` has been defined.
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply cedar-policy/CHANGELOG.md]: Added locally, thanks!

**Review by john-h-kastner-aws — APPROVED** [review]
